src/tests/test_take_until.cpp

```diff
@@ -0,0 +1,107 @@
+        auto mock = mock_observer<int>{};
+
+        rpp::source::create<int>([other_subscriber = subject.get_subscriber()](const auto& subscriber)
+                                 {
+                                    subscriber.on_next(1);
+                                    subscriber.on_next(2);
+
+                                    // Should see a terminate event after this
+                                    other_subscriber.on_next(true);
+
+                                    subscriber.on_next(3);
+                                 })
+            .take_until(subject.get_observable())
+            .subscribe(mock);
+
+        THEN("should see -1-2-|")
+        {
+            CHECK(mock.get_received_values() == std::vector<int>{1, 2});
+            CHECK(mock.get_on_completed_count() == 1);
+            CHECK(mock.get_on_error_count() == 0);
+        }
+    }
+
+    GIVEN("observable of -1-2-3-| pairs with trigger observable of never")
+    {
+        auto mock = mock_observer<int>{};
+        rpp::source::just(1, 2, 3)
+            .take_until(rpp::source::never<bool>())
+            .subscribe(mock);
+
+        THEN("should see -1-2-3-|")
+        {
+            CHECK(mock.get_received_values() == std::vector<int>{1, 2, 3});
+            CHECK(mock.get_on_completed_count() == 1);
+            CHECK(mock.get_on_error_count() == 0);
+        }
+    }
+
+    GIVEN("observable of -1-| pairs with trigger observable of -1-|")
+    {
+        auto mock = mock_observer<int>{};
+        rpp::source::just(1)
+            .take_until(rpp::source::just(1))
+            .subscribe(mock);
+
+        THEN("should see -| because take_until is subscribed first and it also mirrors the trigger observable 's completed event")
+        {
+            CHECK(mock.get_received_values().empty());
+            CHECK(mock.get_on_completed_count() == 1);
+            CHECK(mock.get_on_error_count() == 0);
+        }
+    }
+
+    GIVEN("observable of -1-| pairs with trigger observable of -x")
+    {
+        auto mock = mock_observer<int>{};
+
+        rpp::source::just(1)
+            .take_until(rpp::source::error<bool>(std::make_exception_ptr(std::runtime_error{""})))
+            .subscribe(mock);
+
+        THEN("should see -x because take_until also mirrors the trigger observable 's error event")
+        {
+            CHECK(mock.get_received_values().empty());
+            CHECK(mock.get_on_completed_count() == 0);
+            CHECK(mock.get_on_error_count() == 1);
+        }
+    }
+}
+
+SCENARIO("take_until handles race condition", "[take_until]")
+{
+    GIVEN("trigger value is emitted when source on_next() is still emitting.")
+    {
+        auto mock = mock_observer<size_t>{};
+
+        rpp::source::interval(std::chrono::seconds{1}, rpp::schedulers::trampoline{})
+            .take_until(rpp::source::interval(std::chrono::seconds{1}, rpp::schedulers::trampoline{}))
+            .subscribe(mock);
+    }
```

Comment:
There is no any race-condition =) Because everything happens at the same thread =)

Reply:
Oh my bad, I accidentally commit this and I know this doesn't simulate race-condition at all.